tidy3d/components/geometry/base.py

```diff
@@ -1091,23 +1067,30 @@ def from_gds(
         :class:`Geometry`
             Geometries created from the 2D data.
         """
+        try:
```

Comment:
Yeah it's a bit more confusing I agree.

So the idea was to consolidate the way we do lazy imports into a single methodology with the goal of optimizing that single function for timing performance.

The objective of this implementation was to let importlib sort out the namespace for us since that way we don't have to track library state variables so much. However I could, for example, refactor this to load the library using the `importlib.import_module` method which returns the library and this would remove the need for the `try-except` statements. The idea was to implement the same logic without having to manage the library state through ancilliary variables. However, can improve this implementation if we want to make it cleaner - but it again comes down to how do we do this best when having importlib manage the right library state. 

Let me see if there's other ways to do this (and also relevant for arch 3.0)

Reply:
the way I read it is the difference between 1 and 2 below.

```py
def foo(x):
    have_package_a = check_module_import("A")
    if have_package_a and isinstance(x, TypeA):
        import A
        A.do_something(x)
    have_package_b = check_module_import("B")
    if have_package_b and isinstance(x, TypeB):
        import B
        B.do_something(x)
```

vs.

```py
def foo(x):
    if isinstance(x, TypeA):
        try:
            import A
            A.do_something(x)
        except ImportError:
            ...
            
    if isinstance(x, TypeB):
        try:
            import B
            B.do_something(x)
        except ImportError:
            ...
```
I find 1 to be easier to reason about and avoids try/except. I think the functionality / logic is the same, just that the try except is hidden in the "check_module_import()" in 1? 

> So the idea was to consolidate the way we do lazy imports into a single methodology with the goal of optimizing that single function for timing performance.

could you explain what this single methodology would look like in the context of `foo` above? I'm having trouble seeing it.

Reply:
Yeah sounds good, it makes sense this way, had not thought about it in this implementation - and this will remove the `try-except` blocks so I'll restructure this.

Reply:
So I think maybe the try statements are essential because of the way the `TypeA` and `TypeB` is defined as `gdspy.Cell` or `gdstk.Cell` type which means that the import of the library has to be done beforehand.

```
gdstk_available = check_import("gdstk")
if isinstance(gds_cell, gdstk.Cell):
    import gdstk

    if isinstance(gds_cell, gdstk.Cell):
        gds_loader_fn = Geometry.load_gds_vertic
```

Reply:
oh right, ok feel free to revert.

Reply:
If you import and test the type only after checking for availability, you can still skip the try/except, can't you?
```python
gdstk_available = check_import("gdstk")
if gdstk_available:
    import gdstk

    if isinstance(gds_cell, gdstk.Cell):
        gds_loader_fn = ...
```

Reply:
Sure sounds good. It may even be closer to the original logic just that on a local functional level. 

I guess my mind was a bit mentally blocked on this. Will restructure now to remove the try-excepts this way and it's cleaner to read. 

```
gdstk_available = check_import("gdstk")
if gdstk_available:
    import gdstk

    if isinstance(gds_cell, gdstk.Cell):
        gds_loader_fn = ...
        
gdspy_available = check_import("gdspy")
if gdspy_available:
    import gdspy

    if isinstance(gds_cell, gdspy.Cell):
        gds_loader_fn = ...

if not gdstk_available and gdspy_available:
    ...
```